doc/designs/quic-design/quic-api.md

```diff
@@ -0,0 +1,653 @@
+  non-partial-write `SSL_write` operation spanning multiple `SSL_write` calls,
+  this operation is aborted and partial write mode begins immediately.
+
+- `SSL_MODE_ACCEPT_MOVING_WRITE_BUFFER`: Implemented.
+
+- `SSL_MODE_AUTO_RETRY`: TBD.
+
+- `SSL_MODE_RELEASE_BUFFERS`: Ignored. This is an optimization and if it has
+  any sensible semantic correspondence to QUIC, this can be considered later.
+
+- `SSL_MODE_SEND_FALLBACK_SCSV`: TBD: Either ignore or fail if the client
+  attempts to set this prior to handshake. The latter is probably safer.
+
+  TBD: What if the client attempts to set this post handshake? Ignore it?
+
+- `SSL_MODE_ASYNC`: TBD.
+
+### New APIs
+
+TBD: Should any of these be implemented as ctrls rather than actual functions?
+
+#### `SSL_tick`
+
+| Semantics | `SSL_get_error` | Can Tick? | CSHL          |
+| --------- | ------------- | --------- | ------------- |
+| New       | Never         | Yes       | CS            |
+
+Advances the QUIC state machine to the extent feasible, potentially performing
+network I/O. Also compatible with DTLSv1 and supercedes `DTLSv1_handle_timeout`
+for all use cases.
+
+TBD: Should we just map this to DTLS_CTRL_HANDLE_TIMEOUT internally (and maybe
+alias the CTRL #define)?
+
+#### `SSL_get_tick_timeout`
+
+| Semantics | `SSL_get_error` | Can Tick? | CSHL          |
+| --------- | ------------- | --------- | ------------- |
+| New       | Never         | No        | CS            |
+
+Gets the time until the QUIC state machine next wants to receive a timeout
+event, if any.
+
+This is similar to the existing `DTLSv1_get_timeout` function, but it is not
+specific to DTLSv1. It is also usable for DTLSv1 and can become a
+protocol-agnostic API for this purpose, superceding `DTLSv1_get_timeout` for all
+use cases.
+
+The design is similar to that of `DTLSv1_get_timeout` and uses a `struct
+timeval`. However, this function represents an infinite timeout (i.e., no
+timeout) using `tv_sec == -1`, whereas `DTLSv1_get_timeout` represents an
+infinite timeout using a 0 return value, which does not allow a failure
+condition to be distinguished.
+
+TBD: Should we just map this to DTLS_CTRL_GET_TIMEOUT internally (and maybe
+alias the CTRL #define)?
+
+#### `SSL_set_blocking_mode`, `SSL_get_blocking_mode`
+
+| Semantics | `SSL_get_error` | Can Tick? | CSHL          |
+| --------- | ------------- | --------- | ------------- |
+| New       | Never         | No        | CS            |
+
+Turns blocking mode on or off. This is necessary because up until now libssl has
+operated in blocking or non-blocking mode automatically as an emergent
+consequence of whether the underlying network socket is blocking. Since we are
+proposing to use only non-blocking I/O internally, use of blocking semantics at
+the application level must be explicitly configured.
+
+Use on stream objects: It may be feasible to implement this such that different
+QUIC stream SSL objects can have different settings for this option.
```

Comment:
Yikes! Is that really a good idea?

Reply:
Not really a big deal since we emulate blocking semantics via a relatively trivial adaptation layer. From our perspective it makes no difference. If there's some reason it's hazardous we could choose to prohibit it, but I don't see the issue myself.